src/mutable-prim-array-types.hs

```diff
@@ -31,74 +35,106 @@ unsafeCopy ::
+    let toBytes cnt# = cnt# *# (sizeOf# (undefined :: a))
+     in primitive_ $
+            copyMutableByteArray#
+                src# (toBytes soff#) dst# (toBytes doff#) (toBytes n#)
 
+-------------------------------------------------------------------------------
+-- Length
+-------------------------------------------------------------------------------
+
+-- XXX rename to byteCount?
+{-# INLINE byteLength #-}
+byteLength :: Array s a -> Int
+byteLength (Array arr#) = I# (sizeofMutableByteArray# arr#)
+
+-- XXX Rename length to elemCount so that there is no confusion bout what it
+-- means.
+--
+-- XXX Since size of 'a' is statically known, we can replace `quot` with shift
+-- when it is power of 2. Though it may not matter unless length is used too
+-- often.
+--
 {-# INLINE length #-}
-length ::
-       forall s a. Prim a
-    => Array s a
-    -> Int
-length (Array arr#) =
-    I# (quotInt# (sizeofMutableByteArray# arr#) (sizeOf# (undefined :: a)))
+length :: forall s a. Prim a => Array s a -> Int
+length arr = byteLength arr `quot` (sizeOf (undefined :: a))
 
-{-# INLINE spliceTwo #-}
-spliceTwo ::
+-------------------------------------------------------------------------------
+-- Random Access
+-------------------------------------------------------------------------------
+
+-- XXX Rename to unsafeReadIndex
+-- Check correctness
+{-# INLINE unsafeIndexM #-}
+unsafeIndexM ::
        forall m a. (PrimMonad m, Prim a)
     => Array (PrimState m) a
-    -> Array (PrimState m) a
-    -> m (Array (PrimState m) a)
-spliceTwo a1 a2 = do
-    a3 <- resizeArray a1 (l1 + l2)
-    unsafeCopy a2 0 a3 l1 l2
-    return a3
-  where
-    l1 = length a1
-    l2 = length a2
+    -> Int
+    -> m a
+unsafeIndexM (Array arr#) (I# i#) =
+    -- XXX XXX why not use readByteArray#
+    primitive $ \s# ->
+        case indexByteArray# (unsafeCoerce# arr#) i# of
+            a -> (# s#, a #)
 
+-- XXX rename to unsafeWriteIndex?
 {-# INLINE writeArray #-}
 writeArray ::
        forall m a. (PrimMonad m, Prim a)
     => Array (PrimState m) a -- ^ array
     -> Int -- ^ index
     -> a -- ^ element
     -> m ()
-writeArray (Array arr#) (I# i#) x
-  = primitive_ (writeByteArray# arr# i# x)
+writeArray (Array arr#) (I# i#) x = primitive_ (writeByteArray# arr# i# x)
+
+-------------------------------------------------------------------------------
+-- Construction
+-------------------------------------------------------------------------------
 
+-- Note: We do not store the actual length of the array in the Array
+-- constructor. Therefore, for "length" API to work correctly we need to match
+-- the ByteArray length with the used length by shrinking it.
+--
+-- However, it may be expensive to always shrink the array. We may want to
+-- shrink only if significant space is being wasted. If we want to do that then
+-- we will have to store the used length separately. Or does GHC take care of
+-- that?
+--
```

Comment:
GHC does not take care of it. Also, I don't think in place shrinking is expensive though. Will check. Storing the length separately looks very ugly. I would not recommend it if it isn't required.

Reply:
Does GHC actually reallocate and copy the array even if it is to be shrunk by one byte? Or does it just adjust the length in some cases?

Reply:
The shrinking is always inplace.